### ISSUE TITLE
(Android 3.0) Animated GIFs don't animate.

### ISSUE THREAD

**Issue by DmitriyZaitsev** [koral--/android-gif-drawable#135]
Reproducible on Android 3.0 Galaxy Tab 7.0 Plus.

**Comment by koral--** [comment]
I could not reproduce that on API 11 emulator. I've tried `GifImageView` and `TextView` with `GifDrawable` as a background and both animations are working.

**Comment by DmitriyZaitsev** [comment]
I'll have that device on Monday and give you more info.

**Comment by Doctoror** [comment]
Actually I could reproduce it.
Checkout my project sample from branch **dev**
https://github.com/Doctoror/ImageFactory/tree/dev

There is a GridView with identical resources decoded and the first one is not animating until you scroll down to the bottom and up again.

Launch this on hdpi device 480x800 or so to reproduce.

If you move creating the drawables from onCreate with postDelayed with 2000 delay or so the issue is not occurring.

I had exactly the same bug when creating AnimationDrawable from code and the issue fixed when I tried overriding onDraw() of the target View like

````
@Override
protected void onDraw(@NonNull final Canvas c) {
    super.onDraw();
    final Drawable drawable = getDrawable();
    if (drawable.getCallback() != this) {
        drawable.setCallback(this);
    }
}
```

Though I see you don't use Drawable's scheduleSelf but it behaves exactly the same with normal AnimationDrawable.

Play around with my sample and maybe you can get idea on what's going on.

**Comment by koral--** [comment]
@Doctoror reason is that drawable at index 0 is associated with more than one view. In my case it has been called 5 times with 2 different instances of `GifImageView` (without any scrolling) while for other indexes it has been called once. Moreover view from last call is not visible one. That's why item at index 0 is not animating: the same drawable instance is set on visible view and then on invisible one which is not redrawn. I don't know (yet) if it is intended behavior of Adapter but it is very interesting.

Simpler workaround in your case (without subclassing) can be something like this:
```
Index: sample/src/main/java/com/doctoror/imagefactory/sample/DemoActivity.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- sample/src/main/java/com/doctoror/imagefactory/sample/DemoActivity.java	(revision aa1ebff81dbcab3928589e045100a8bf2be92d79)
+++ sample/src/main/java/com/doctoror/imagefactory/sample/DemoActivity.java	(revision )
@@ -117,7 +117,9 @@
                 imageView = (ImageView) convertView.getTag();
             }
 
+            if (getItem(position).getCallback() == null) {
-            imageView.setImageDrawable(getItem(position));
+                imageView.setImageDrawable(getItem(position));
+            }
             return convertView;
         }
     }
```
Remember that `#getCallback()` is available on API 11+.

**Comment by Doctoror** [comment]
So, you're setting the `Drawable` to the `ImageView` only if the `Callback` is not set. Which doesn't look like a great idea to do. Since when you scroll down the `Drawable`'s callback won't be `null` thus the ImageView's drawable won't be changed. You will not notice since in the demo all `Drawable`s look the same.

I checked that and I see `getView()` is called in weird sequence:
0
1
2
3
4
5
0
0

It's probably related to layout pass but not sure how.

If you move setting Adapter to` Activity` `onWindowFocusChanged()` you will get a different more predictable sequence

```java
@Override
    public void onWindowFocusChanged(final boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        final GridView grid = (GridView) findViewById(R.id.activity_demo_grid);
        grid.setAdapter(new DemoAdapter(this, generateImageInfo()));
    }
```

0
0
1
2
3
4
5

I still don't get it what's with calling `getView(0)` two times but Animation is working for `View` at index 0 this time.

So I see phew workarounds for `AdapterView`s.
Either store a Set of WeakReference<Callback> instead of single Callback for the GifDrawable or call setCallback(this) in all `View`s that support GifDrawable.

**Comment by Doctoror** [comment]
I've created a GifDrawable2 that stores a Set of WeakReferences and overridden GifImageView and GifTextView to manage GifDrawable2 callback list.

Check it out on **adapterview_bug_fix** branch.
https://github.com/Doctoror/ImageFactory/tree/adapterview_bug_fix

**Comment by koral--** [comment]
@Doctoror IMO the better solution is something like:
```
for (GifImageView view:allGifImageViewInGridView){
   if(view.isOnScreen())
      view.getDrawable().setCallback(view);
}
```
that traversal may be performed eg. in `onGlobalLayout()` of `OnGlobalLayoutListener` set in `getView()`.
Maybe `View#getDrawingRect()` can be used to check if view is on screen.

**Comment by Doctoror** [comment]
Dont forget getBackground()` and `getCompoundDrawables()` for `TextView`.

In this case it's easier to override `onAttachedToWindow()` calling `setCallback()` on all View's `Drawables`. Not sure it will work but I'll give it a try. If it works it's probably the ultimate solution.

**Comment by dangalg** [comment]
I didn't quite understand the solution? should I use doctors GifDrawable2? will that solve my problem?

**Comment by Doctoror** [comment]
@dangalg that solves the problem with first AdapterView's first View that don't animate, but it's a bit nasty solution.

@koral-- are you gonna do something about this?

**Comment by dangalg** [comment]
I am thinking of a simple hack solution for this issue. How about if we make the two first items identical and then just remove the first one... I will try this out later.

**Comment by koral--** [comment]
@Doctoror yes, I am planing to include some solution in library.

**Comment by koral--** [comment]
@dangalg @Doctoror I have new idea, it should work for more cases than solution proposed until now. Basically invalidate all views which hosts given drawable. It will work for all `View` subclasses (eg. layouts with `GifDrawable` as background, not necessarily in adapters), does not require modifications in existing library classes and is not using any dirty hacks.
Example based on ImageFactory: https://github.com/koral--/ImageFactory/blob/multi_drawable/sample/src/main/java/com/doctoror/imagefactory/sample/DemoActivity.java#L126
Do you see any drawbacks?

**Comment by dangalg** [comment]
I added the MultiCallback class to my app and this is my code for getview:

 @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            View v = convertView;
            ViewHolder holder; // to reference the child views for later actions

            if (v == null) {
                LayoutInflater vi =
                        (LayoutInflater) getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
                v = vi.inflate(R.layout.item_package, null);
                // cache view fields into the holder
                holder = new ViewHolder();
                holder.emuticon_animated_gif = (GifImageView) v.findViewById(R.id.emuticon_animated_gif);
                // associate the holder with the view for later lookup

                v.setTag(holder);
                if(position == 0){
                    Log.d(TAG, String.valueOf(position));
                    Log.d(TAG, String.valueOf(v));
                    v.invalidate();
                }
            }
            else {
                // view already exists, get the holder instance from the view
                holder = (ViewHolder) v.getTag();
                if(position == 0){
                    v.invalidate();
                }
            }
            // no local variables with findViewById here
            final GifDrawable pack = items.get(position);
            if (pack != null) {
                if (holder.emuticon_animated_gif != null) {
                    if(position == 0) {
                        Drawable.Callback callback = pack.getCallback();
                        holder.emuticon_animated_gif.setImageDrawable(items.get(position));
                        if (!(callback instanceof MultiCallback)) {
                            callback = new MultiCallback();
                        }
                        ((MultiCallback) callback).addView(holder.emuticon_animated_gif);
                        pack.setCallback(callback);
                    }
                    else{
                        holder.emuticon_animated_gif.setImageDrawable(items.get(position));
                    }
                }
            }

            // use holder.nameText where you were
            // using the local variable nameText before
            return v;
        }
    }

    // somewhere else in your class definition
    static class ViewHolder {
        GifImageView emuticon_animated_gif;
    }

And now sometimes it plays and sometimes it doesn't. It seems unstable. am I doing something wrong?

**Comment by Doctoror** [comment]
@koral-- 

II modified it a little. See the comments in code. Other than that, it seems okay at first look.

```java
    private static final class DemoAdapter extends BaseAdapter2<Drawable> {

        private final MultiCallback mMultiCallback = new MultiCallback();
        
        private DemoAdapter(@NonNull final Context context,
                @Nullable final List<Drawable> items) {
            super(context, items);
        }

        @Override
        public View getView(final int position, View convertView, final ViewGroup parent) {
            final ImageView imageView;
            if (convertView == null) {
                convertView = getLayoutInflater().inflate(R.layout.grid_item_demo, parent, false);
                imageView = (ImageView) convertView.findViewById(R.id.image);
                convertView.setTag(imageView);
            } else {
                imageView = (ImageView) convertView.getTag();
            }
            
            // getCallback() is available only at API level 11. Since we support lower, avoid it.
            // Instead, create single MultiCallback and set it every time. 

            final Drawable item = getItem(position);
            imageView.setImageDrawable(item);
            mMultiCallback.addView(imageView);
            item.setCallback(mMultiCallback);

            return convertView;
        }
    }

    public static class MultiCallback implements Drawable.Callback {

        private final Set<WeakReference<View>> mViewSet = new LinkedHashSet<>();

        @Override
        public void invalidateDrawable(final Drawable who) {
            final Iterator<WeakReference<View>> iterator = mViewSet.iterator();
            while (iterator.hasNext()) {
                final View view = iterator.next().get();
                if (view != null) {
                    view.invalidateDrawable(who);
                } else {
                    // Always remove null references to reduce Set size
                    iterator.remove();
                }
            }
        }

        @Override
        public void scheduleDrawable(final Drawable who, final Runnable what, final long when) {
            final Iterator<WeakReference<View>> iterator = mViewSet.iterator();
            while (iterator.hasNext()) {
                final View view = iterator.next().get();
                if (view != null) {
                    view.scheduleDrawable(who, what, when);
                } else {
                    // Always remove null references to reduce Set size
                    iterator.remove();
                }
            }
        }

        @Override
        public void unscheduleDrawable(final Drawable who, final Runnable what) {
            final Iterator<WeakReference<View>> iterator = mViewSet.iterator();
            while (iterator.hasNext()) {
                final View view = iterator.next().get();
                if (view != null) {
                    view.unscheduleDrawable(who, what);
                } else {
                    // Always remove null references to reduce Set size
                    iterator.remove();
                }
            }
        }

        public void addView(final View view) {
            // mViewSet.add(new WeakReference<>(view)); Wrong. WeakReference does not
            // override equals so you will end up with lots of references pointing to the same
            // instance
            final Iterator<WeakReference<View>> iterator = mViewSet.iterator();
            while (iterator.hasNext()) {
                final View item = iterator.next().get();
                if (item == null) {
                    // Always remove null references to reduce Set size
                    iterator.remove();
                } else if (item == view) {
                    // Return when found or if loop end reached, add it.
                    return;
                }
            }
            mViewSet.add(new WeakReference<>(view));
        }

        public void removeView(final View view) {
            final Iterator<WeakReference<View>> iterator = mViewSet.iterator();
            while (iterator.hasNext()) {
                final View item = iterator.next().get();
                if (item == null || item == view) {
                    // Always remove null references to reduce Set size
                    iterator.remove();
                }
            }
        }
    }
```

**Comment by dangalg** [comment]
Where does baseadapter2 come from?

**Comment by dangalg** [comment]
Ok used your code and tested and it works very very well!! thanks so much!

**Comment by koral--** [comment]
@Doctoror I've found and fixed 2 issues:
1. animation of first item had stopped if grid is scrolled down eg. about a half of height of the item but only first time. I don't know (yet) exact reason, but fix (workaround?) is to replace `View#invalidateDrawable()` with `View#invalidate()`.
2. If `MultiCallback#addView()` would be used by two (or more) parallel threads the following scenario might be possible:
* 1st thread reaches end of the while loop and is preempted
* 2nd thread reaches end of the while loop and is preempted
* 1st thread adds `WeakReference` to the given view
* 2nd thread adds `WeakReference` to the same view
That can be solved eg. by making `addView` synchronized or by subclassing `WeakReference`, override `equals()` and check for referents equality there, so `mViewSet` won't contain multiple references to the same 'View`. I used second solution.

**Comment by Doctoror** [comment]
@koral-- LinkedHashSet is not thread-safe so I think you must add a lock on all methods that can modify it. Note that synchronizing methods will not work here if you can call `MultiCallback#addView()` and `MultiCallback#removeView()` asynchronously.

Also, if I'm correct, `Collections.synchronizedSet() will not work here since the Iterator will not be synchronized.

**Comment by koral--** [comment]
Right, some thread-safe collection should be used. I think `CopyOnWriteArrayList` will be better there than synchronized blocks because synchronization affects performance and invalidation is normally much more frequent than collection modifications. `CopyOnWriteArrayList#get()` does not need to be synchronized because copy is made on each list modification (`addIfAbsent()` and `remove()` methods are synchronized).
Additionally returning from `addView()` from a middle of a loop can cause that references with null referents may left at the end of a collection. Removing references to nulls is not a main feature there however.

I just commited version with `CopyOnWriteArrayList` and it is also available in latest snapshot.

**Comment by koral--** [comment]
I've done some research and noticed that issue 1. from https://github.com/koral--/android-gif-drawable/issues/135#issuecomment-101084596 is caused by the fact that one of the views for item at index 0 is reused (new row appears at the bottom). From this point only drawable for view at index 0 passed to `Callback#invalidateDrawable()` is different from that returned by `ImageView#getDrawable()` so `ImageView#verifyDrawable()` returns always false.

One of the solutions is to create `GifDrawable`s using builder with `GifDrawableBuilder#setRenderingTriggeredOnDraw(false)` called.

Moreover snippet in https://github.com/koral--/android-gif-drawable/issues/135#issuecomment-100668514 uses one `MultiCallback` instance for all grid items, so all visible items are invalidated and redrawn if any (not only its) `Drawable` needs to be invalidated.

With mentioned `setRenderingTriggeredOnDraw(false)` separate callbacks for each item can be used (only that for item at index 0 will have more than 1 `View` associated).

So to sum up, the most elegant solution is now that `MultiCallback` should use `View#invalidateDrawable()` and (in case of adapters) `GifDrawable` should be constructed using builder with `setRenderingTriggeredOnDraw(false)`. I'll add README section about that.

@Doctoror do you think it is enough, or have some better idea/something to add?

**Comment by Doctoror** [comment]
@koral-- 

You are correct. The issue with single MultiCallback in snippet https://github.com/koral--/android-gif-drawable/issues/135#issuecomment-100668514 can be easily fixed by creating a `Drawable`-to-`MultiCallback` `Map`.

```java
    private static final class DemoAdapter extends BaseAdapter2<Drawable> {

        private final Map<Drawable, MultiCallback> mCallbackMap = new HashMap<>();

        private DemoAdapter(@NonNull final Context context,
                @Nullable final List<Drawable> items) {
            super(context, items);
        }

        @Override
        public View getView(final int position, View convertView, final ViewGroup parent) {
            final ImageView imageView;
            if (convertView == null) {
                convertView = getLayoutInflater().inflate(R.layout.grid_item_demo,
                        parent, false);
                imageView = (ImageView) convertView.findViewById(R.id.image);
                convertView.setTag(imageView);
            } else {
                imageView = (ImageView) convertView.getTag();
            }

            // getCallback() is available only at API level 11. Since we support lower
            // we should avoid it. Instead, create a single MultiCallback and set it
            // every time.

            final Drawable item = getItem(position);
            MultiCallback callback = mCallbackMap.get(item);
            if (callback == null) {
                callback = new MultiCallback();
                mCallbackMap.put(item, callback);
            }
            imageView.setImageDrawable(item);
            callback.addView(imageView);
            item.setCallback(callback);

            return convertView;
        }
    }
```

I have no comments regarding `setRenderingTriggeredOnDraw(false)` since I'm not sure how the resulting Drawable behaves with and without it and don't have any spare time at the moment to go deeper on this.

**Comment by koral--** [comment]
Information about `MultiCallback` added to README. Closing original issue due to lack of information, can be reopened if some info is provided.
